src/zql/context/replicache-context.ts

```diff
@@ -0,0 +1,149 @@
+import {InvariantViolation} from '../error/invariant-violation.js';
+import {compareEntityFields} from '../ivm/compare.js';
+
+export function makeReplicacheContext(rep: Replicache) {
+  const materialite = new Materialite();
+  const sourceStore = new ReplicacheSourceStore(rep, materialite);
+
+  return {
+    materialite,
+    getSource: <T extends Entity>(
+      name: string,
+      ordering?: [string[], 'asc' | 'desc'],
+    ) => sourceStore.getSource(name, ordering) as Source<T>,
+  };
+}
+
+/**
+ * Forwards Replicache changes to Materialite sources so they
+ * can be fed into any queries that may exist.
+ *
+ * Maintains derived orderings of sources as well.
+ *
+ * If someone runs a query that has an order-by we need to scan the entire collection
+ * in order to sort it.
+ * To save future work, we save the result of that sort and keep it up to date.
+ *
+ * This helps:
+ * 1. When revisting old queries that were sorted or paging through results
+ * 2. When many queries are sorted by the same field
+ * 3. When joining sources on a field that we have pre-sorted
+ *
+ * And shares the work between queries.
+ */
+class ReplicacheSourceStore {
+  readonly #rep: Replicache;
+  readonly #materialite: Materialite;
+  readonly #sources = new Map<string, ReplicacheSource>();
+
+  constructor(rep: Replicache, materialite: Materialite) {
+    this.#rep = rep;
+    this.#materialite = materialite;
+  }
+
+  getSource(name: string, ordering?: [string[], 'asc' | 'desc']) {
+    let source = this.#sources.get(name);
+    if (source === undefined) {
+      source = new ReplicacheSource(this.#rep, this.#materialite, name);
+      this.#sources.set(name, source);
+    }
+
+    return source.get(ordering);
+  }
+}
+
+class ReplicacheSource {
+  readonly #materialite;
+  readonly #sources: Map<string, Source<Entity>> = new Map();
+  readonly #canonicalSource: MutableSetSource<Entity>;
+
+  constructor(rep: Replicache, materialite: Materialite, name: string) {
+    this.#canonicalSource =
+      materialite.newSetSource<Entity>(canonicalComparator);
+    this.#materialite = materialite;
+    rep.experimentalWatch(this.#onReplicacheDiff, {
+      prefix: `${name}/`,
+      initialValuesInFirstDiff: true,
+    });
+  }
+
+  #onReplicacheDiff = (changes: ExperimentalNoIndexDiff) => {
+    this.#materialite.tx(() => {
+      for (const diff of changes) {
+        if ('oldValue' in diff) {
+          // This lookup of old was due to seeing `oldValue` values that did not match
+          // what was stored on the client when working on Repliear.
+          // `oldValue` not matching what is on the client is problematic
+          // if there are derived sources with orderings that depend on fields outside the id.
+          // E.g.,
+          // If a derived source is sorted by `modified_time`
+          // and the `oldValue` provided has a different `modified_time` we'll fail to remove
+          // the correct value from the derived source.
```

Comment:
Dax has reported something like this to me, but he didn't ever have time to make a reproduction.

Looking at the code, neither Greg nor I can imagine any way this can happen. If you have a reproduction let's fix it.

Reply:
Thank you very much for noticing/flagging this @arv - totally missed it.

Reply:
This was several months ago that I saw the behavior. I'll see if I can put together a consistent repro though.

Reply:
I wonder if this could happen when changes are batched. Perhaps there is some bug where multiple updates to the same key get smooshed together as one `experimentalWatch` update and the `oldValue` for some reason is not the one that's currently in Replicache, but one of the other ones that got smooshed together. Just a random guess.

Reply:
Correction. watch is still experimental but I would still like to figure this out.